pkg/router/route_group_test.go

```diff
@@ -21,120 +23,51 @@ import (
 	require.NotNil(t, rg)
 	require.Equal(t, DefaultRouteGroupConfig(), rg.cfg)
 }
 
 func TestRouteGroup_Close(t *testing.T) {
-	keys := snettest.GenKeyPairs(2)
-
-	pk1 := keys[0].PK
-	pk2 := keys[1].PK
-
-	// create test env
-	nEnv := snettest.NewEnv(t, keys, []string{stcp.Type})
-	defer nEnv.Teardown()
-
-	tpDisc := transport.NewDiscoveryMock()
-	tpKeys := snettest.GenKeyPairs(2)
-
-	m1, m2, tp1, tp2, err := transport.CreateTransportPair(tpDisc, tpKeys, nEnv, stcp.Type)
-	require.NoError(t, err)
-	require.NotNil(t, tp1)
-	require.NotNil(t, tp2)
-	require.NotNil(t, tp1.Entry)
-	require.NotNil(t, tp2.Entry)
-
-	rg0 := createRouteGroup()
-	rg1 := createRouteGroup()
-
-	// reserve FWD and CNSM IDs for r0.
-	r0RtIDs, err := rg0.rt.ReserveKeys(2)
-	require.NoError(t, err)
-
-	// reserve FWD and CNSM IDs for r1.
-	r1RtIDs, err := rg1.rt.ReserveKeys(2)
-	require.NoError(t, err)
-
-	r0FwdRule := routing.ForwardRule(ruleKeepAlive, r0RtIDs[0], r1RtIDs[1], tp1.Entry.ID, pk2, pk1, 0, 0)
-	r0CnsmRule := routing.ConsumeRule(ruleKeepAlive, r0RtIDs[1], pk1, pk2, 0, 0)
+	rg1, rg2, m1, m2, teardown := setupEnv(t)
 
-	err = rg0.rt.SaveRule(r0FwdRule)
-	require.NoError(t, err)
-	err = rg0.rt.SaveRule(r0CnsmRule)
-	require.NoError(t, err)
-
-	r1FwdRule := routing.ForwardRule(ruleKeepAlive, r1RtIDs[0], r0RtIDs[1], tp2.Entry.ID, pk1, pk2, 0, 0)
-	r1CnsmRule := routing.ConsumeRule(ruleKeepAlive, r1RtIDs[1], pk2, pk1, 0, 0)
-
-	err = rg1.rt.SaveRule(r1FwdRule)
-	require.NoError(t, err)
-	err = rg1.rt.SaveRule(r1CnsmRule)
-	require.NoError(t, err)
-
-	r0FwdRtDesc := r0FwdRule.RouteDescriptor()
-	rg0.desc = r0FwdRtDesc.Invert()
-	rg0.tps = append(rg0.tps, tp1)
-	rg0.fwd = append(rg0.fwd, r0FwdRule)
-
-	r1FwdRtDesc := r1FwdRule.RouteDescriptor()
-	rg1.desc = r1FwdRtDesc.Invert()
-	rg1.tps = append(rg1.tps, tp2)
-	rg1.fwd = append(rg1.fwd, r1FwdRule)
-
-	// push close packet from transport to route group
-	go func() {
-		packet, err := m1.ReadPacket()
-		if err != nil {
-			panic(err)
-		}
-
-		if packet.Type() != routing.ClosePacket {
-			panic("wrong packet type")
-		}
-
-		if err := rg0.handleClosePacket(routing.CloseCode(packet.Payload()[0])); err != nil {
-			panic(err)
-		}
-	}()
+	ctx, cancel := context.WithCancel(context.Background())
 
 	// push close packet from transport to route group
-	go func() {
-		packet, err := m2.ReadPacket()
-		if err != nil {
-			panic(err)
-		}
-
-		if packet.Type() != routing.ClosePacket {
-			panic("wrong packet type")
-		}
-
-		if err := rg1.handleClosePacket(routing.CloseCode(packet.Payload()[0])); err != nil {
-			panic(err)
-		}
-	}()
+	go pushPackets(ctx, m2, rg2)
+	go pushPackets(ctx, m1, rg1)
 
-	err = rg0.Close()
+	err := rg1.Close()
 	require.NoError(t, err)
-	require.True(t, rg0.isClosed())
-	require.True(t, rg1.isRemoteClosed())
+	require.True(t, rg1.isClosed())
+	require.True(t, rg2.isRemoteClosed())
 	// rg1 should be done (not getting any new data, returning `io.EOF` on further reads)
 	// but not closed
```

Comment:
Should be specify:
```golang
// rg1 should be closed for reads (returning `io.EOF` on further reads), but not on writes.
```
Is this the intended behaviour?

Reply:
@evanlinjin yep. `io.EOF` is only returned when the connection is broken (we test this via `isRemoteClosed`) and no data left on the wire (we check the `readCh` for it). this was the only way to make route group behave like the usual connection, and also it seems like the most correct one to me